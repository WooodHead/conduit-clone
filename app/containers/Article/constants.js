/*
 *
 * Article constants
 *
 */

export const DEFAULT_ACTION = 'app/Article/DEFAULT_ACTION';
export const FETCH_ARTICLE = 'app/Article/FETCH_ARTICLE';
export const ARTICLE_LOADED = 'app/Article/ARTICLE_LOADED';
export const LOAD_ARTICLE_ERROR = 'app/Article/LOAD_ARTICLE_ERROR';
