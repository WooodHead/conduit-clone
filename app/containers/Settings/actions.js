/*
 *
 * Settings actions
 *
 */

import {
  UPDATE_SETTINGS,
  UPDATE_FETCHING,
  UPDATE_USER_LOADED,
  UPDATE_USER_ERROR,
} from './constants';

export function updateSettings(form) {
  return {
    type: UPDATE_SETTINGS,
    form,
  };
}

export function updateFetching() {
  return {
    type: UPDATE_FETCHING,
  };
}

export function updateUserLoaded(user) {
  return {
    type: UPDATE_USER_LOADED,
    user,
  };
}

export function updateUserError(err) {
  return {
    type: UPDATE_USER_ERROR,
    err,
  };
}
