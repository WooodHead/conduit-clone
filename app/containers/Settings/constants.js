/*
 *
 * Settings constants
 *
 */

export const DEFAULT_ACTION = 'app/Settings/DEFAULT_ACTION';
export const UPDATE_SETTINGS = 'app/Settings/UPDATE_SETTINGS';
export const UPDATE_FETCHING = 'app/Settings/UPDATE_FETCHING';
export const UPDATE_USER_LOADED = 'app/Settings/UPDATE_USER_LOADED';
export const UPDATE_USER_ERROR = 'app/Settings/UPDATE_USER_ERROR';
