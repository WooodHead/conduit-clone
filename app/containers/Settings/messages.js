/*
 * Settings Messages
 *
 * This contains all the text for the Settings component.
 */
import { defineMessages } from 'react-intl';

export default defineMessages({
  logout: {
    id: 'app.containers.Settings.logout',
    defaultMessage: 'Logout',
  },
});
