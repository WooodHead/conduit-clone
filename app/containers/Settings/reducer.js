/*
 *
 * Settings reducer
 *
 */

import { fromJS } from 'immutable';
import {
  UPDATE_FETCHING,
  UPDATE_USER_LOADED,
} from './constants';


const initialState = fromJS({
  fetching: false,
  success: false,
});

function settingsReducer(state = initialState, action) {
  switch (action.type) {
    case UPDATE_FETCHING:
      return state.set('fetching', true);
    case UPDATE_USER_LOADED:
      return state
        .set('fetching', false)
        .set('success', true);
    default:
      return state;
  }
}

export default settingsReducer;
