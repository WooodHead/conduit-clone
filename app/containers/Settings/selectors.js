import { createSelector } from 'reselect';

/**
 * Direct selector to the settings state domain
 */
const makeSelectSettingsDomain = (state) => state.get('settings');

/**
 * Default selector used by Settings
 */
const makeSelectFetching = () => createSelector(
  makeSelectSettingsDomain,
  (substate) => substate.get('fetching')
);

const makeSelectSuccess = () => createSelector(
  makeSelectSettingsDomain,
  (substate) => substate.get('success')
);

export {
  makeSelectSettingsDomain,
  makeSelectFetching,
  makeSelectSuccess,
};
