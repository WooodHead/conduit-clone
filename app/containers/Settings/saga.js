import { takeLatest, put, call } from 'redux-saga/effects';
import { UPDATE_SETTINGS } from 'containers/Settings/constants';

import {
  updateFetching,
  updateUserLoaded,
  updateUserError,
} from 'containers/Settings/actions';
import api from '../../utils/api';

export function* fetchUpdateSettingValues(params) {
  yield put(updateFetching());
  try {
    const payload = yield call(api.Auth.updateSettings, params);
    yield put(updateUserLoaded(payload.user));
  } catch (err) {
    yield put(updateUserError(err));
  }
}

export default function* onUpdateSettingsSaga() {
  yield takeLatest(UPDATE_SETTINGS, fetchUpdateSettingValues);
}
