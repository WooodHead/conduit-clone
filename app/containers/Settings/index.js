/**
 *
 * Settings
 *
 */

import React from 'react';
import PropTypes from 'prop-types';
import { push } from 'react-router-redux';
import { connect } from 'react-redux';
import { Helmet } from 'react-helmet';
import { FormattedMessage } from 'react-intl';
import { createStructuredSelector } from 'reselect';
import { compose } from 'redux';

import { Container, Button } from 'reactstrap';

import injectSaga from 'utils/injectSaga';
import injectReducer from 'utils/injectReducer';
import makeSelectSettings from './selectors';
import reducer from './reducer';
import saga from './saga';
import messages from './messages';

import { destroy } from '../../auth';

export class Settings extends React.PureComponent { // eslint-disable-line react/prefer-stateless-function
  constructor() {
    super();
    this.onLogout = this.onLogout.bind(this);
  }
  onLogout() {
    destroy();
    this.props.onLogout();
  }
  render() {
    return (
      <Container>
        <Helmet>
          <title>Settings</title>
          <meta name="description" content="Description of Settings" />
        </Helmet>
        <Button outline color="primary" onClick={this.onLogout}>
          <FormattedMessage {...messages.logout} />
        </Button>
      </Container>
    );
  }
}

Settings.propTypes = {
  onLogout: PropTypes.func.isRequired,
};

const mapStateToProps = createStructuredSelector({
  settings: makeSelectSettings(),
});

function mapDispatchToProps(dispatch) {
  return {
    onLogout: () => dispatch(push('/')),
  };
}

const withConnect = connect(mapStateToProps, mapDispatchToProps);

const withReducer = injectReducer({ key: 'settings', reducer });
const withSaga = injectSaga({ key: 'settings', saga });

export default compose(
  withReducer,
  withSaga,
  withConnect,
)(Settings);
