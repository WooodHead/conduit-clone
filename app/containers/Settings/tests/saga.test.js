/**
 * Test  sagas
 */

/* eslint-disable redux-saga/yield-effects */
import { takeLatest, put } from 'redux-saga/effects';
import { UPDATE_SETTINGS } from 'containers/Settings/constants';
import {
  updateUserError,
} from '../actions';
import onUpdateSettingsSaga, { fetchUpdateSettingValues } from '../saga';

describe('defaultSaga Saga', () => {
  let fetchUpdateSettingValuesGenerator;
  beforeEach(() => {
    const user = {
      username: 'ausername',
      email: 'email@web.com',
      password: 'sample password',
    };
    fetchUpdateSettingValuesGenerator = fetchUpdateSettingValues(user);
    let putDescriptor = fetchUpdateSettingValuesGenerator.next().value;
    expect(putDescriptor).toMatchSnapshot();

    const callDescriptor = fetchUpdateSettingValuesGenerator.next().value;
    expect(callDescriptor).toMatchSnapshot();

    putDescriptor = fetchUpdateSettingValuesGenerator.next({ user }).value;
    expect(putDescriptor).toMatchSnapshot();
  });

  it('should dispatch the updateUserError action if it requests throws an error', () => {
    const putDescriptor = fetchUpdateSettingValuesGenerator.throw('Error').value;
    expect(putDescriptor).toEqual(put(updateUserError('Error')));
  });
});

describe('onUpdateSettingsSaga Saga', () => {
  const defaultSaga = onUpdateSettingsSaga();
  it('should start task to watch for UPDATE_SETTINGS action', () => {
    const takeLatestDescriptor = defaultSaga.next().value;
    expect(takeLatestDescriptor).toEqual(takeLatest(UPDATE_SETTINGS, fetchUpdateSettingValues));
  });
});
