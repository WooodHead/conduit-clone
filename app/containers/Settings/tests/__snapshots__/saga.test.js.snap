// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`defaultSaga Saga should dispatch the updateUserError action if it requests throws an error 1`] = `
Object {
  "@@redux-saga/IO": true,
  "PUT": Object {
    "action": Object {
      "type": "app/Settings/UPDATE_FETCHING",
    },
    "channel": null,
  },
}
`;

exports[`defaultSaga Saga should dispatch the updateUserError action if it requests throws an error 2`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [
      Object {
        "email": "email@web.com",
        "password": "sample password",
        "username": "ausername",
      },
    ],
    "context": null,
    "fn": [Function],
  },
}
`;

exports[`defaultSaga Saga should dispatch the updateUserError action if it requests throws an error 3`] = `
Object {
  "@@redux-saga/IO": true,
  "PUT": Object {
    "action": Object {
      "type": "app/Settings/UPDATE_USER_LOADED",
      "user": Object {
        "email": "email@web.com",
        "password": "sample password",
        "username": "ausername",
      },
    },
    "channel": null,
  },
}
`;
